metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/usage.py

```diff
@@ -143,6 +152,140 @@ def bigquery_audit_metadata_query_template(
     return textwrap.dedent(query)
 
 
+class BigQueryUsageState:
+    read_events: FileBackedDict[ReadEvent]
+    query_events: FileBackedDict[QueryEvent]
+    column_accesses: FileBackedDict[Tuple[str, str]]
+
+    def __init__(self, config: BigQueryV2Config):
+        self.conn = ConnectionWrapper()
+        self.read_events = FileBackedDict[ReadEvent](
+            shared_connection=self.conn,
+            tablename="read_events",
+            extra_columns={
+                "resource": lambda e: str(e.resource),
+                "name": lambda e: e.jobName,
+                "timestamp": lambda e: get_time_bucket(
+                    e.timestamp, config.bucket_duration
+                ),
+                "user": lambda e: e.actor_email,
+            },
+        )
+        # Keyed by job_name
+        self.query_events = FileBackedDict[QueryEvent](
+            shared_connection=self.conn,
+            tablename="query_events",
+            extra_columns={
+                "query": lambda e: e.query,
+                "is_read": lambda e: int(e.statementType in READ_STATEMENT_TYPES),
+            },
+        )
+        # Created just to store column accesses in sqlite for JOIN
+        self.column_accesses = FileBackedDict[Tuple[str, str]](
+            shared_connection=self.conn,
+            tablename="column_accesses",
+            extra_columns={"read_event": lambda p: p[0], "field": lambda p: p[1]},
+        )
+
+    def __enter__(self) -> "BigQueryUsageState":
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_val: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
+    ) -> None:
+        self.read_events.close()
+        self.query_events.close()
+        self.column_accesses.close()
+        self.conn.close()
+
+    def standalone_events(self) -> Iterable[AuditEvent]:
+        for read_event in self.read_events.values():
+            query_event = (
+                self.query_events.get(read_event.jobName)
+                if read_event.jobName
+                else None
+            )
+            yield AuditEvent(read_event=read_event, query_event=query_event)
+        for _, query_event in self.query_events.filtered_items("NOT is_read"):
+            yield AuditEvent(query_event=query_event)
+
+    def statistics_buckets(self) -> Iterable[Tuple[str, str]]:
+        query = f"""
+        SELECT DISTINCT
+          timestamp,
+          resource
+        FROM {self.read_events.tablename}
+        ORDER BY timestamp
+        """
```

Comment:
this is cool.
So we always write to SQLite irrespective of whether data size (FileBackedDict) is small or large ?